### PR TITLE
feat(las): support laz 1.4

### PR DESCRIPTION
**Pull request by alekzvik**
This is WIP 

Changes so far:
- Add support for LAZ v1.4
- Change decompressor for LAZ chunks from laz-perf to laz-rs
- Update examples to include LAZ 1.4 file (took from PDAL/data, missing attribution)
- More point formats. All 10 from LAS 1.4 spec are covered.
- A new example needs to be uploaded to vis.gl/data with proper attribution - visgl/deck.gl-data#38

Notes:
On `laz-perf` -> `laz-rs` change: `laz-rs` has a smaller bundle size and was easier for me to understand. `laz-rs` wasm bindings repo does not have a license though, I created an issue there - laz-rs/laz-rs-wasm#2. 

Current blockers:

- [x] data file PR - visgl/deck.gl-data#38
- [ ] laz-rs wasm bindings license PR - laz-rs/laz-rs-wasm#2

### CODE REVIEW

**Comment by alekzvik** [comment]
Hello, this is my first PR to loaders.gl. I glanced through contributor guidelines, but if I missed anything, please let me know.

**Review by ibgreen — COMMENTED** [review]
This is looking quite promising. I wasn't aware there was a WASM-transpiled Rust parser for LAZ.

File: modules/las/src/las-loader.ts [hunk 1]
```diff
@@ -23,7 +22,7 @@ export type LASLoaderOptions = LoaderOptions & {
     /** Override the URL to the worker bundle (by default loads from unpkg.com) */
     workerUrl?: string;
   };
-  onProgress?: (progress: number) => void;
+  onProgress?: Function;
```
[the remark this answers is not in the document]
Comment (alekzvik) [reply modules/las/src/las-loader.ts]: That's an open question for me now.
onProgress callbacks are rarely used across other loaders. And LAS one [is actually using a different signature than I specified here](https://github.com/visgl/loaders.gl/blob/3d9fed050eabdc0812ddf2f4d5fb9914a34ee0c2/modules/las/src/lib/parse-las.ts#L127-L135):
```
    const meshBatch = {
      ...lasMesh,
      header: {
        vertexCount: lasHeader.totalRead
      },
      progress: lasHeader.totalRead / lasHeader.totalToRead
    };

    options?.onProgress?.(meshBatch);
```
The only other part in loaders that used onProgress callbacks is fetch-progress - [link](https://github.com/visgl/loaders.gl/blob/3d9fed050eabdc0812ddf2f4d5fb9914a34ee0c2/modules/core/src/lib/progress/fetch-progress.ts#L68)

And I like that fetch-progress does not pass so much data around and is centered around one use case - reporting progress. I think if we want progress callback in LAS, we can follow the same approach and model the signature after the [ProgressEvent](https://developer.mozilla.org/en-US/docs/Web/API/ProgressEvent):
```
type ProgressUpdate = {
  total?: number;
  loaded: number;
  lengthComputable: boolean;
}
```
What do you think? Does it make sense?

**Comment by alekzvik** [comment]
> This is looking quite promising.

Thanks!
> I wasn't aware there was a WASM-transpiled Rust parser for LAZ.

Yes, I did not find any usage of the WASM bindings in the open source libraries, but the underlying `las-rz` is used across many las readers in the rust ecosystem: `las-rs` & `copc_rs`

**Comment by ibgreen** [comment]
> * Add support for LAZ v1.4

That is awesome

> * Change decompressor for LAZ chunks from laz-perf to laz-rs

Since backwards compatiblitiy is important for a well-used framework like loaders.gl, what do you think about supporting both for a little while?

We could have a 
```ts
export LASPlasioLoader = ...;
export LASRustLoader` = ;
export const LASLoader=LASRustLoader;
```

This way the old loader is available for a while until we gain confidence. 

> * More point formats. All 10 from LAS 1.4 spec are covered.
> * A new example needs to be uploaded to vis.gl/data with proper attribution - [feat: Add LAZ 1.4 sample file to use on point cloud examples deck.gl-data#38](https://github.com/visgl/deck.gl-data/pull/38)
> 


> 
> Current blockers:
> 
> * data file PR - [feat: Add LAZ 1.4 sample file to use on point cloud examples deck.gl-data#38](https://github.com/visgl/deck.gl-data/pull/38)
> * Update examples to include LAZ 1.4 file (took from PDAL/data, missing attribution)

Please update READMEs here and in deck.gl-data to specify the source of this file, ideally the license if you have it. 

> * laz-rs wasm bindings license PR - [License laz-rs/laz-rs-wasm#2](https://github.com/laz-rs/laz-rs-wasm/issues/2)

Yes. Thanks for reaching out to the author. We really need this to be in place before we merge.

**Comment by connormanning** [comment]
Has the decompression performance between these two libraries been investigated?  We use `laz-perf` in [copc.js](https://github.com/connormanning/copc.js), and if the Rust one is faster by any significant degree I would look into swapping it out.  On the flip side it would be nice to know if it happened to be any *slower* for consideration of this PR.

I assume they are largely equivalent but it would be nice to know for sure.  I could potentially set up a benchmark based on real-world files via their usage in [viewer.copc.io](https://viewer.copc.io), but I was curious if anyone had already spent time looking into this.

**Comment by ibgreen** [comment]
@connormanning Thanks for weighing in. 

- Yes I was suggesting to @alekzvik that maybe we start by adding this as a new parallel LAS loader (instead of outright replacing the old one), then we can do some comparative benchmarking and get comfortable with its backwards compatibility etc.

- I do have a question. I saw in COPC.js that you use laz-perf 0.0.5 with some 1.4 extension code, whereas in loaders.gl we use v0.0.6 which clearly doesn't support v1.4. What is the situation there?

**Comment by connormanning** [comment]
> whereas in loaders.gl we use v0.0.6 which clearly doesn't support v1.4.

In what way is it not supported exactly?  As far as I know, 1.4 support landed in the source code [2021](https://github.com/hobuinc/laz-perf/commit/dfec0f6ddc958d615aeb274487c60219dcec9d6e), although I didn't gain control of the `laz-perf` package on npm until 2022 when I published [`0.0.5`](https://www.npmjs.com/package/laz-perf?activeTab=versions) (`0.0.6` was just a minor tweak and I think is functionally identical).  Prior to that you would have had to build it yourself.

It is true that the API exposed by laz-perf is very low level (so I fully get @alekzvik's comment about the Rust one being "easier for me to understand"), but [here](https://github.com/connormanning/copc.js/blob/master/src/las/point-data.ts) we have abstractions for both chunk-level decoding (used for COPC, which is required to be 1.4) as well as file-level decoding.  The [unit tests](https://github.com/connormanning/copc.js/blob/master/src/las/point-data.test.ts) for both of these APIs test against 1.4 files and succeed with both `0.0.5` and `0.0.6`, which I just upgraded/tested locally.  These two APIs are used 

Maybe some of this abstraction could be hoisted up into `laz-perf` itself.  Since that repository is already a bit complex due to the mixture of C++ and JS, I did it downstream and left `laz-perf` simply an exact representation of the low-level Emscripten bindings.  Note that copc.js is MIT licensed so you can steal/modify these wrappers as you wish.

If you have an example I can reproduce that demonstrates the failure to support 1.4 that you are seeing, I can look into it.

**Comment by ibgreen** [comment]
> In what way is it not supported exactly? As far as I know, 1.4 support landed in the source code..

I apologize, I honestly had no idea that laz-perf already supports 1.4, if so I would have attempted to make use of that a long time ago. My info was obviously outdated and based on earlier discussions such as #591 which made it sound like that code was going nowhere, so I suppose I just stopped paying attention.

So the situation seems to have improved greatly, we now have at least two options for 1.4 support. 

May the best parser win :)

**Comment by connormanning** [comment]
That makes sense.  The broader/updated context, considering [this](https://github.com/visgl/loaders.gl/issues/591#issuecomment-573420978) comment from 2020 about how 1.4 support was *not* on the immediate roadmap at the time, is that the following year we developed the COPC format and defined it as being a 1.4 file, leading to those newer developments.  This was for simplicity so reader implementations wouldn't have to support both 1.2 and 1.4.

From an API level I am sure the Rust version is cleaner, since laz-perf requires you to understand things like the Emscripten heap, and to take care not to leak Emscripten module resources.  I'm not opposed to upstreaming some abstractions for it, but so far there haven't been requests for that.  I would guess that people tend to grab the wrapped functionality via `import { Las } from 'copc'` rather than using laz-perf directly.  Especially since there are a lot of other useful LAS abstractions there as well.

If you are not using the other LAS utilities (header/extra-bytes parsing, dimension "view" extraction, etc.) from copc.js and you just want to decompress LAZ chunks/files, then the Rust version is probably simpler.

**Comment by ibgreen** [comment]
Ok that does answer some other questions. If we building the LAZLoader on copc.js was a reasonable option.

It seems to me that the overhead of the remaining code is not that big compared to the laz parser itself, so probably not an issue.

If we want to support COPC, then we need your LAZ loader anyway right, (unless we modify COPC to make the las loader pluggable).

> From an API level I am sure the Rust version is cleaner, since laz-perf requires you to understand things like the Emscripten heap, and to take care not to leak Emscripten module resources. 

Interesting... Aren't they both WASM and subject to the same level of messyness? Or do rust modules compile more cleanly?

**Comment by connormanning** [comment]
> Aren't they both WASM and subject to the same level of "messyness"?

To an extent yes, but there are basically 2 levels where you can decide on APIs for these things.

The first is the API at the WASM boundary itself.  For laz-perf, that looks like [this](https://github.com/hobuinc/laz-perf/blob/master/cpp/emscripten/laz-perf.cpp).  Different laz-perf libraries might have completely different interfaces here, for example the choice to expose the decompression as a per-point thing [here](https://github.com/hobuinc/laz-perf/blob/2b68700d33bf981d257d6101d96714620c64019c/cpp/emscripten/laz-perf.cpp#L25) and [here](https://github.com/hobuinc/laz-perf/blob/master/cpp/emscripten/laz-perf.cpp#L63) is a developer choice and could be different.  This particular API choice was made in 2014, for context.  The details have changed a bit since then but generally the same model remained.  It could be a higher level API.

The second is what you expose via JS.  Emscripten generates WASM with the above API exposed from C++ land, as well as emitting JS bootstrap code which invokes the WASM.  [`createLazPerf`](https://github.com/hobuinc/laz-perf/blob/master/js/src/index.ts) here is just a raw re-export of the Emcripten binding output, and I added TS [types](https://github.com/hobuinc/laz-perf/blob/master/js/src/laz-perf.d.ts) to it but left it untouched.  When you `import { createLazPerf }` from `laz-perf`, this is what you are importing.  Here is where you might add JS-level abstractions like the layer you see in `copc.js`, or other forms of utilities that access the Emscripten bindings without exposing them explicitly.  For example, importing `laz-perf` could *theoretically* give you some form of [these two functions](https://github.com/connormanning/copc.js/blob/master/src/las/point-data.ts#L68-L105), encapsulating the init/heap-allocate/loop/destroy logic rather than giving you the Emscripten module and making *you* do that as the importer.  So similar to the above, this portion could be a higher level API.  And of course at this level, since you're in JS now, you could add other APIs/utilities/wrappers in addition to the existing low-level exports, like the other LAS stuff you see in copc.js.

I don't really know how the Rust one works as I haven't looked at it.  But if they've made different choices for their API and exposed things at a higher level, then it would be simpler from the perspective of an importer.  For my part, I just left `laz-perf` alone as a representation of the existing low-level APIs and put my abstractions in `copc.js`.

**Comment by alekzvik** [comment]
A few comments from my side.

> I do have a question. I saw in COPC.js that you use laz-perf 0.0.5 with some 1.4 extension code, whereas in loaders.gl we use v0.0.6 which clearly doesn't support v1.4. What is the situation there?

loaders.gl used an older version of `laz-perf` that was compiled on 2020-07-23, [see here](https://github.com/visgl/loaders.gl/blob/master/modules/las/src/lib/libs/laz-perf.ts#L5) and this version did not have laz 1.4 support. `laz-perf` version 0.0.6 was added to package.json without actually importing laz-perf code from the npm module. So this might have caused some confusion.

> Since backwards compatiblitiy is important for a well-used framework like loaders.gl, what do you think about supporting both for a little while?

I will take a look at this and see how awkward it will be. But overall, it makes sense. 

> Has the decompression performance between these two libraries been investigated?

I have not. But it's a great idea and worth investigating. I agree with @ibgreen that supporting two loaders side by side will most probably make it easier. 

About the `copc.js`. I haven't looked inside yet but I like what @connormanning explained here, and maybe it's worth looking into either importing or copying some parts of it. 

And one final point about `laz-perf` vs `laz-rs`. The API that is currently exposed in `laz-rs` is stateful. The underlying rust code keeps the "pointer" to what has already been "decompressed", so it might be worse and more cumbersome in the case of batch parsing. I can also take a look at laz-rs and see if there are additional APIs that are there but not yet exposed to WASM bindings.

**Comment by ibgreen** [comment]
@alekzvik I landed your PR and a follow-up which adds the previous laz-perf based loader in parallel.
Now that we have both loaders in place we could do some comparative testing perhaps using the existing benchmarking setup in loaders.gl (`yarn test bench`, `yarn test bench-browser`)

Moved the discussion to https://github.com/visgl/loaders.gl/discussions/3170

**Comment by alekzvik** [comment]
Final note: laz-rs-wasm bindings added an MIT license to the repo

**Comment by ibgreen** [comment]
> Final note: laz-rs-wasm bindings added an MIT license to the repo

Yes thanks, I saw that which made me decide it was OK to merge.

**Pull request merged**